core/src/main/scala/org/apache/spark/ExecutorAllocationClient.scala

```diff
@@ -81,6 +81,35 @@ private[spark] trait ExecutorAllocationClient {
     countFailures: Boolean,
     force: Boolean = false): Seq[String]
 
+  /**
+   * Request that the cluster manager decommission the specified executors.
+   * Default implementation delegates to kill, scheduler must override
+   * if it supports graceful decommissioning.
+   *
+   * @param executorIds identifiers of executors to decommission
+   * @param adjustTargetNumExecutors whether the target number of executors will be adjusted down
+   *                                 after these executors have been decommissioned.
+   * @return the ids of the executors acknowledged by the cluster manager to be removed.
+   */
+  def decommissionExecutors(
+    executorIds: Seq[String],
+    adjustTargetNumExecutors: Boolean): Seq[String]
+
+
+  /**
+   * Request that the cluster manager decommission the specified executors.
+   * Default implementation delegates to kill, scheduler must override
+   * if it supports graceful decommissioning.
+   *
+   * @param executorIds identifiers of executors to decommission
+   * @return whether the request is acknowledged by the cluster manager.
+   */
+  def decommissionExecutor(executorId: String): Boolean = {
+    val decommissionedExecutors = decommissionExecutors(Seq(executorId),
+      adjustTargetNumExecutors = true)
```

Comment:
Why is `adjustTargetNumExecutors` defaulting to true here ? This would mean that all schedulers would try to replinish the executor when asked to `DecommissionExecutor(...)` -- for example by the Master or when an executor gets a SIGPWR. 

I think it shouldn't be the default -- it should atleast be configurable. It only makes sense to have `adjustTargetNumExecutors=true` when called from `org.apache.spark.streaming.scheduler.ExecutorAllocationManager#killExecutor` (ie when it is truly called from dynamic allocation codepath and we have decided that we want to replinish the executor).

Reply:
If you look above there is a configurable call. This matches how `killExecutor` is implemented down on line 124.

Reply:
Can you please point me to where is the configurable call ? I don't see a config check in the code paths that call this method.

It's fine for killExecutor to unconditionally adjust the target number of executors because it is only called in the dynamic allocation codepath, but decommissionExecutor would be called from many other codepaths as well (for example when the driver gets a DecommissionExecutor message) -- and thus I think it should just assume that it should replenish the executor.

Reply:
Look on line 95 of this file. I think we should match the semantics of `killExecutor` as much as possible. If there's a place where we don't want it we can use `decommissionExecutors`

Reply:
Hmm, Should we rename decommissionExecutor (singular) to decommissionAndKillExecutor to reflect its purpose better ? It would be too easy to confuse it with decommissionExecutors (on line 95 of this file which allows to not replenish the target number of executors).

Do you want to make the change to the callers of decommissionExecutor in this PR and switch them to `decommissioExecutors(Seq(executorId), false)` instead. The ones I am most concerned about are:

- The handling of message DecommissionExecutor (both sync and async variants) in CoarseGrainedSchedulerBackend
- StandaloneSchedulerBackend.executorDecommissioned

In both the above cases, I think we may not _always_ want replenishing. For example, in the standalone case, when the Worker gets a SIGPWR -- do we want to replenish the executors on the remaining workers (ie oversubscribe the remaining workers) ? Similarly when an executor gets a SIGPWR, do we want to put that load on the remaining executors ? I think the answer to both should be NO unless we are doing a dynamic allocation.

Personally I am fine with any choice of naming here as long as the semantics are not silently changed under the cover, as is the case presently.

Reply:
It's a new function, what are we changing?

Reply:
ExecutorAllocationClient is a base class of CoarseGrainedSchedulerBackend. We moved decommissionExecutor from the latter class to the former and as such it is not a new function. Since CoarseGrainedSchedulerBackend no longer overrides decommissionExecutor, ExecutorAllocationClient.decommissionExecutor will be called when CoarseGrainedSchedulerBackend gets a DecommissionExecutor message -- and the semantics of that codepath have been changed to unconditionally impose adjustTargetNumExecutors=true.

Reply:
cool, I'll update the previous calls to `decommissioExecutor`